Shared/Supporting Files/Models/OnDemandResource.swift

```diff
@@ -35,54 +35,59 @@ final class OnDemandResource: ObservableObject {
     /// The progress of the on-demand resource request.
     var progress: Progress { request.progress }
     
+    /// A Boolean value indicating whether a resource request can be initiated.
+    var isDownloadable: Bool {
+        requestState != .inProgress && requestState != .downloaded
+    }
+    
     /// The current state of the on-demand resource request.
-    @Published private(set) var requestState: RequestState = .notStarted
+    private(set) var requestState: RequestState?
     
     /// The error occurred in downloading resources.
-    @Published private(set) var error: (any Error)?
+    private(set) var error: (any Error)?
     
     /// The on-demand resource request.
     private let request: NSBundleResourceRequest
     
-    /// A set of cancellable instances for the request progress subscription.
-    private var cancellables: Set<AnyCancellable> = []
+    /// The name of the sample associated with the resource request.
+    let sampleName: String
     
-    /// Initializes a request with a set of Resource Tags.
-    init(tags: Set<String>) {
-        request = NSBundleResourceRequest(tags: tags)
+    /// Initializes a request with a sample.
+    init(sample: Sample) {
+        self.sampleName = sample.name
+        request = NSBundleResourceRequest(tags: [sample.nameInUpperCamelCase])
         request.loadingPriority = NSBundleResourceRequestLoadingPriorityUrgent
-        request.progress
-            .publisher(for: \.fractionCompleted, options: .new)
-            .receive(on: DispatchQueue.main)
-            .map { .inProgress($0) }
-            .sink { [weak self] in self?.requestState = $0 }
-            .store(in: &cancellables)
+    }
+    
+    /// Sets up the `requestState` by checking if the resource is already on device.
+    func setUp() async {
+        guard requestState == nil else { return }
+        
+        let isResourceAvailable = await request.conditionallyBeginAccessingResources()
+        requestState = isResourceAvailable ? .downloaded : .notStarted
     }
     
     /// Cancels the on-demand resource request.
     func cancel() {
         progress.cancel()
-        cancellables.removeAll()
         request.endAccessingResources()
         requestState = .cancelled
     }
     
     /// Starts the on-demand resource request.
     func download() async {
-        // Initiates download when it is not being/already downloaded.
-        // Checks if the resource is already on device.
-        let isResourceAvailable = await request.conditionallyBeginAccessingResources()
-        if isResourceAvailable {
+        guard isDownloadable else { return }
+        
+        requestState = .inProgress
+        do {
+            try await request.beginAccessingResources()
```

Comment:
It is worth pointing out this, from the [doc](https://developer.apple.com/documentation/foundation/nsbundleresourcerequest/endaccessingresources()#Discussion) for `endAccessingResources()` (emphasis mine):

> Call this method as soon as you have finished using the tags managed by this request. **If needed, this method will be called by the system when the resource request object is deallocated.**

For the resource downloader, this means that any of the downloaded resources could be purged as soon as the downloader dismisses and the resources are deallocated. I'd say that warrants a comment and perhaps even a message in the UI that downloads could be purged by the system at any time.